### PR TITLE
Weak delegates refactoring

### PR DESCRIPTION
**Pull request by TimofeyBurak**
### Description

Refactor WeakAction/WeakFunc inheritance

### API Changes

Added:
 - `abstract class WeakDelegate`
 - `bool IExecuteWithObject.IsAlive` // used in Messenger
 - `string IExecuteWithObject.MethodName` // used in Messenger

Changed:
 - `class WeakAction<T> : WeakAction  => class WeakAction<T> : WeakDelegate<Action<T>>`
 - `class WeakFunc<TIn, TOut> : WeakFunc<TOut> => class WeakFunc<TIn, TOut> : WeakDelegate<Func<TIn, TOut>>`
 - `WeakReference WeakDelegate.Reference => WeakReference WeakDelegate.CustomTargetReference`

Removed:
 - `WeakAction` default constructor
 - `WeakFunc` default constructor

### Platforms Affected

- Core (all platforms)

### Behavioral/Visual Changes

- `WeakAction<T>` now is not a subclass of `WeakAction`, therefore doesn't support parameterless `Execute()` method;
- `WeakFunc<TIn, TOut>` now is not a subclass of `WeakFunc<TOut>`, therefore doesn't support parameterless `Execute()` method;
- Property `Target` now returns either Custom Target (if provided in constructor and alive) or Delegate's Target (if alive) for all WeakDelegate subclasses;
- `IsAlive` now returns false if Custom Target not alive, even for static delegates;
- `Execute()` now does nothing if Custom Target not alive, even for static delegates;

### Before/After Screenshots

Not applicable

### PR Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/Softeq/XToolkit.WhiteLabel/blob/master/.github/CONTRIBUTING.md) document
- [x] My code follows the [code styles](https://github.com/Softeq/dotnet-guidelines)
- [x] Targets the correct branch
- [x] Tests are passing (or failures are unrelated)

### CODE REVIEW

File: Softeq.XToolkit.Common.Tests/WeakTests/Utils/DisposableCleanupAction.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Developed by Softeq Development Corporation
+// http://www.softeq.com
+
+using System;
+
+namespace Softeq.XToolkit.Common.Tests.WeakTests.Utils
+{
+    public sealed class DisposableCleanupAction : IDisposable
```
[the remark this answers is not in the document]
Comment (TimofeyBurak) [reply Softeq.XToolkit.Common.Tests/WeakTests/Utils/DisposableCleanupAction.cs]: Thank you for pointing out! Fixed

File: Softeq.XToolkit.Common.Tests/WeakTests/WeakDelegatesTests.DataProviders.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        public static IEnumerable<object[]> WeakActionInputParameters => TypesGenerator().Select(x => new[] { x });
+        public static IEnumerable<object[]> WeakFuncOutputParameters => TypesGenerator().Select(x => new[] { x });
+        public static IEnumerable<object[]> WeakFuncInputOutputParameters =>
+            TypesGenerator().SelectMany(t => TypesGenerator(), (x, y) => new[] { x, y });
+
+        private static IEnumerable<object> TypesGenerator()
+        {
+            yield return 42;    // int
+            yield return 42.0;  // double
```
[the remark this answers is not in the document]
Comment (TimofeyBurak) [reply Softeq.XToolkit.Common.Tests/WeakTests/WeakDelegatesTests.DataProviders.cs]: Updated

File: Softeq.XToolkit.Common.Tests/WeakTests/WeakDelegatesTests.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
+﻿// Developed by Softeq Development Corporation
+// http://www.softeq.com
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+#nullable disable
```
Comment (wcoder) [inline Softeq.XToolkit.Common.Tests/WeakTests/WeakDelegatesTests.cs:7]: Can be removed, because disabled on the top level (for all Tests projects)

Comment (TimofeyBurak) [reply Softeq.XToolkit.Common.Tests/WeakTests/WeakDelegatesTests.cs]: Fixed

**Pull request merged**